File: packages/python/hashrope/sliding.py
"""
Sliding window hash rope for bounded-memory streaming hash computation.

Adapted from UHC's SlidingRopeState (Definition 10, Theorem 11).
Provides a generic sliding window over a hash rope without any
compression-format dependencies.

The window maintains a prefix hash and a bounded rope, evicting old
data as new data arrives. At any point, the full-sequence hash
H(T[0..pos-1]) can be computed in O(1) from the prefix hash and
the window rope hash.

Invariant I_slide (holds after every operation):
    (1) h_prefix = H(T[0 .. l_prefix - 1])
    (2) r_window represents T[l_prefix .. pos - 1]
    (3) rope_len(r_window) = pos - l_prefix <= W
    (4) H(T[0..pos-1]) = h_prefix * x^(pos - l_prefix) + rope_hash(r_window)
"""

from __future__ import annotations

from hashrope.polynomial_hash import PolynomialHash, MERSENNE_61
from hashrope.rope import (
    Leaf,
    Node,
    rope_concat,
    rope_split,
    rope_repeat,
    rope_len,
    rope_hash,
)


class SlidingWindow:
    """
    Bounded-memory sliding window over a hash rope.

    Maintains a window of at most W = d_max + m_max bytes as a rope,
    plus a prefix hash covering all evicted bytes. Supports two
    operations:

    - append_bytes(data): append literal bytes
    - append_copy(offset, length): append a copy from within the window

    At any point, current_hash() returns H(T[0..pos-1]) in O(1).

    Parameters
    ----------
    d_max : int
        Maximum copy-back distance. The window retains at least d_max
        bytes after eviction so that future copies can find their source.
    m_max : int
        Maximum copy length. Together with d_max, determines the
        maximum window extent W = d_max + m_max.
    prime : int
        Mersenne prime for hash arithmetic. Default: 2^61 - 1.
    base : int
        Hash base. Default: 131.
    """

    __slots__ = (
        "h", "d_max", "m_max", "W",
        "h_prefix", "l_prefix", "r_window", "pos",
    )

    def __init__(
        self,
        d_max: int = 32768,
        m_max: int = 258,
        prime: int = MERSENNE_61,
        base: int = 131,
    ) -> None:
        if d_max < 1:
            raise ValueError(f"d_max must be >= 1, got {d_max}")
        if m_max < 1:
            raise ValueError(f"m_max must be >= 1, got {m_max}")

        self.h = PolynomialHash(prime=prime, base=base)
        self.d_max = d_max
        self.m_max = m_max
        self.W = d_max + m_max

        # State (Definition 10)
        self.h_prefix: int = 0       # H(T[0..l_prefix-1])
        self.l_prefix: int = 0       # evicted byte count
        self.r_window: Node = None   # rope for T[l_prefix..pos-1]
        self.pos: int = 0            # total bytes ingested

    @property
    def window_len(self) -> int:
        """Current window size in bytes."""
        return rope_len(self.r_window)

    def current_hash(self) -> int:
        """
        Compute H(T[0..pos-1]) from sliding state via Theorem 1:
            H(T) = H_prefix * x^(window_len) + H(window)
        """
        wl = self.window_len
        wh = rope_hash(self.r_window) if self.r_window else 0
        return self.h.hash_concat(self.h_prefix, wl, wh)

    def final_hash(self) -> int:
        """
        FinalHash() -- Theorem 11.

        H(T[0..N-1]) = H_prefix * x^(R_window.len) + R_window.hash
        """
        return self.current_hash()

    def append_bytes(self, data: bytes) -> None:
        """
        Append literal bytes to the stream.

        Parameters
        ----------
        data : bytes
            One or more bytes to append.
        """
        if not data:
            return
        leaf = Leaf(data, self.h)
        self.r_window = rope_concat(self.r_window, leaf, self.h)
        self.pos += len(data)
        self._evict()

    def append_copy(self, offset: int, length: int) -> None:
        """
        Append a copy-from-window reference.

        Copies `length` bytes starting `offset` bytes back from the
        current end of the window. Handles both non-overlapping
        (offset >= length) and overlapping (offset < length) cases.

        Parameters
        ----------
        offset : int
            Distance back from the current end of the window.
            Must be <= pos and <= window_len.
        length : int
            Number of bytes to copy. For overlapping copies
            (offset < length), the pattern of length `offset`
            is repeated as needed.

        Raises
        ------
        ValueError
            If offset exceeds decoded length or window size.
        """
        win_len = self.window_len

        if offset > self.pos:
            raise ValueError(
                f"Invalid copy: offset {offset} exceeds "
                f"decoded length {self.pos}"
            )

        if offset > win_len:
            raise ValueError(
                f"Copy offset {offset} exceeds window size {win_len}. "
                f"Increase d_max (currently {self.d_max}) to at least {offset}."
            )

        start = win_len - offset

        if offset >= length:
            # Non-overlapping
            _, tmp = rope_split(self.r_window, start, self.h)
            source, _ = rope_split(tmp, length, self.h)
            self.r_window = rope_concat(self.r_window, source, self.h)
        else:
            # Overlapping: extract pattern of length `offset`, repeat
            _, tmp = rope_split(self.r_window, start, self.h)
            pattern, _ = rope_split(tmp, offset, self.h)

            q, r = divmod(length, offset)
            rep: Node = rope_repeat(pattern, q, self.h) if q >= 1 else None
            if r > 0:
                partial, _ = rope_split(pattern, r, self.h)
                rep = rope_concat(rep, partial, self.h)
            self.r_window = rope_concat(self.r_window, rep, self.h)

        self.pos += length
        self._evict()

    def _evict(self) -> None:
        """
        Evict excess bytes from the window.

        After eviction: rope_len(r_window) <= d_max, ensuring that
        future copies of distance <= d_max find their source.

        Correctness (Lemma 12): The prefix hash update follows from
        Theorem 1 (concatenation composability).
        """
        win_len = rope_len(self.r_window)
        if win_len <= self.W:
            return

        # Evict down to d_max bytes remaining
        excess = win_len - self.d_max
        r_old, r_keep = rope_split(self.r_window, excess, self.h)

        # Fold evicted bytes into h_prefix via Theorem 1
        old_len = rope_len(r_old)
        old_hash = rope_hash(r_old)
        self.h_prefix = self.h.hash_concat(self.h_prefix, old_len, old_hash)

        self.l_prefix += excess
        self.r_window = r_keep

File: packages/rust/src/sliding.rs
//! Sliding window hash rope for bounded-memory streaming hash computation.
//!
//! Provides a generic sliding window over a hash rope without any
//! compression-format dependencies.

use crate::rope::*;

/// Bounded-memory sliding window over a hash rope.
///
/// Maintains a window of at most `W = d_max + m_max` bytes as a rope,
/// plus a prefix hash covering all evicted bytes.
pub struct SlidingWindow {
    arena: Arena,
    d_max: u64,
    pub m_max: u64,
    w: u64,
    h_prefix: u64,
    l_prefix: u64,
    r_window: Node,
    pos: u64,
}

impl SlidingWindow {
    /// Create a new sliding window.
    ///
    /// # Panics
    /// Panics if `d_max < 1` or `m_max < 1`.
    pub fn new(d_max: u64, m_max: u64, prime: u64, base: u64) -> Self {
        assert!(d_max >= 1, "d_max must be >= 1, got {}", d_max);
        assert!(m_max >= 1, "m_max must be >= 1, got {}", m_max);
        Self {
            arena: Arena::with_hash(prime, base),
            d_max,
            m_max,
            w: d_max + m_max,
            h_prefix: 0,
            l_prefix: 0,
            r_window: None,
            pos: 0,
        }
    }

    /// Create with default parameters.
    pub fn default_window() -> Self {
        Self::new(32768, 258, crate::polynomial_hash::MERSENNE_61, 131)
    }

    /// Access the arena.
    pub fn arena(&self) -> &Arena {
        &self.arena
    }

    /// Mutable access to the arena.
    pub fn arena_mut(&mut self) -> &mut Arena {
        &mut self.arena
    }

    /// Current window size in bytes.
    #[inline]
    pub fn window_len(&self) -> u64 {
        self.arena.len(self.r_window)
    }

    /// Total bytes ingested.
    #[inline]
    pub fn pos(&self) -> u64 {
        self.pos
    }

    /// Compute `H(T[0..pos-1])` from sliding state.
    pub fn current_hash(&mut self) -> u64 {
        let wl = self.window_len();
        let wh = self.arena.hash(self.r_window);
        self.arena.hasher_mut().hash_concat(self.h_prefix, wl, wh)
    }

    /// Alias for `current_hash()`.
    pub fn final_hash(&mut self) -> u64 {
        self.current_hash()
    }

    /// Append literal bytes to the stream.
    pub fn append_bytes(&mut self, data: &[u8]) {
        if data.is_empty() {
            return;
        }
        let leaf = self.arena.from_bytes(data);
        self.r_window = self.arena.concat(self.r_window, leaf);
        self.pos += data.len() as u64;
        self.evict();
    }

    /// Append a copy-from-window reference.
    ///
    /// Copies `length` bytes starting `offset` bytes back from the
    /// current end of the window.
    ///
    /// # Panics
    /// Panics if offset exceeds decoded length or window size.
    pub fn append_copy(&mut self, offset: u64, length: u64) {
        let win_len = self.window_len();

        assert!(
            offset <= self.pos,
            "Invalid copy: offset {} exceeds decoded length {}",
            offset, self.pos
        );
        assert!(
            offset <= win_len,
            "Copy offset {} exceeds window size {}. Increase d_max (currently {}) to at least {}.",
            offset, win_len, self.d_max, offset
        );

        let start = win_len - offset;

        if offset >= length {
            // Non-overlapping
            let (_, tmp) = self.arena.split(self.r_window, start);
            let (source, _) = self.arena.split(tmp, length);
            self.r_window = self.arena.concat(self.r_window, source);
        } else {
            // Overlapping: extract pattern of length `offset`, repeat
            let (_, tmp) = self.arena.split(self.r_window, start);
            let (pattern, _) = self.arena.split(tmp, offset);

            let q = length / offset;
            let r = length % offset;

            let mut rep = if q >= 1 {
                self.arena.repeat(pattern, q)
            } else {
                None
            };

            if r > 0 {
                let (partial, _) = self.arena.split(pattern, r);
                rep = self.arena.concat(rep, partial);
            }

            self.r_window = self.arena.concat(self.r_window, rep);
        }

        self.pos += length;
        self.evict();
    }

    fn evict(&mut self) {
        let win_len = self.arena.len(self.r_window);
        if win_len <= self.w {
            return;
        }

        let excess = win_len - self.d_max;
        let (r_old, r_keep) = self.arena.split(self.r_window, excess);

        let old_len = self.arena.len(r_old);
        let old_hash = self.arena.hash(r_old);
        self.h_prefix = self.arena.hasher_mut().hash_concat(self.h_prefix, old_len, old_hash);

        self.l_prefix += excess;
        self.r_window = r_keep;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;
    use alloc::vec::Vec;
    use crate::polynomial_hash::MERSENNE_61;

    #[test]
    fn test_empty() {
        let mut sw = SlidingWindow::default_window();
        assert_eq!(sw.current_hash(), 0);
        assert_eq!(sw.pos(), 0);
    }

    #[test]
    fn test_append_bytes() {
        let mut sw = SlidingWindow::default_window();
        let expected = sw.arena().hash_bytes(b"hello");
        sw.append_bytes(b"hello");
        assert_eq!(sw.current_hash(), expected);
    }

    #[test]
    fn test_append_incremental() {
        let mut sw = SlidingWindow::default_window();
        let expected = sw.arena().hash_bytes(b"hello world");
        sw.append_bytes(b"hello ");
        sw.append_bytes(b"world");
        assert_eq!(sw.current_hash(), expected);
    }

    #[test]
    fn test_non_overlapping_copy() {
        let mut sw = SlidingWindow::default_window();
        let expected = sw.arena().hash_bytes(b"hello worldworld");
        sw.append_bytes(b"hello world");
        sw.append_copy(5, 5);
        assert_eq!(sw.current_hash(), expected);
    }

    #[test]
    fn test_overlapping_copy() {
        let mut sw = SlidingWindow::default_window();
        let expected = sw.arena().hash_bytes(b"abababab");
        sw.append_bytes(b"ab");
        sw.append_copy(2, 6);
        assert_eq!(sw.current_hash(), expected);
    }

    #[test]
    fn test_eviction_preserves_hash() {
        let mut sw = SlidingWindow::new(8, 4, MERSENNE_61, 131);
        let data = b"the quick brown fox jumps over the lazy dog";
        let expected = {
            let mut a = Arena::with_hash(MERSENNE_61, 131);
            a.hash_bytes(data)
        };
        for &byte in data.iter() {
            sw.append_bytes(&[byte]);
        }
        assert_eq!(sw.current_hash(), expected);
    }

    #[test]
    fn test_large_stream() {
        let mut sw = SlidingWindow::new(32, 16, MERSENNE_61, 131);
        let mut full_data = Vec::new();
        for i in 0..50u8 {
            let chunk = vec![i; 10];
            sw.append_bytes(&chunk);
            full_data.extend_from_slice(&chunk);
        }
        let expected = {
            let mut a = Arena::with_hash(MERSENNE_61, 131);
            a.hash_bytes(&full_data)
        };
        assert_eq!(sw.current_hash(), expected);
    }
}
